controllers/webhook.js

```diff
@@ -0,0 +1,302 @@
+        reqAuthenticated = false
+        const webhookSecretSent = req.headers['x-gitlab-token']
+        if (!webhookSecretSent) {
+          // This could be worth logging... unless the endpoint gets hammered with spam.
+          errorsRaised.push('No secret found in the webhook request')
+        } else if (webhookSecretExpected === webhookSecretSent) {
+          /*
+           * Whereas GitHub uses the webhook secret to sign the request body, GitLab does not.
+           * As such, just check that the received secret equals the expected value.
+           */
+          reqAuthenticated = true
+        } else {
+          // This could be worth logging... unless the endpoint gets hammered with spam.
+          errorsRaised.push('Unable to verify authenticity of request')
+        }
+      }
+
+      if (reqAuthenticated) {
+        await _handleMergeRequest(req.params, service, req.body, staticman).catch((errors) => {
+          errorsRaised = errors
+        })
+      }
+    }
+  }
+
+  if (errorsRaised.length > 0) {
+    return Promise.reject(errorsRaised)
+  }
+}
+
+const _calcIsVersion1WhenGitHubAssumed = function (req) {
+  const service = req.params.service
+  const version = req.params.version
+  return (!service && version === '1')
+}
+
+const _verifyGitHubSignature = function (secret, data, signature) {
+  const signedData = 'sha1=' + crypto.createHmac('sha1', secret).update(data).digest('hex')
+  return bufferEq(Buffer.from(signature), Buffer.from(signedData))
+}
+
+const _handleMergeRequest = async function (params, service, data, staticman) {
+  const errors = []
+
+  const ua = config.get('analytics.uaTrackingId')
+    ? require('universal-analytics')(config.get('analytics.uaTrackingId'))
+    : null
+
+  const version = params.version
+  const username = params.username
+  const repository = params.repository
+  const branch = params.branch
+
+  let gitService = null
+  let mergeReqNbr = null
+  if (service === 'github') {
+    gitService = await _buildGitHubService(version, service, username, repository, branch, data)
+    mergeReqNbr = data.number
+  } else if (service === 'gitlab') {
+    gitService = await gitFactory.create('gitlab', {
+      version: version,
+      username: username,
+      repository: repository,
+      branch: branch
+    })
+    mergeReqNbr = data.object_attributes.iid
+  } else {
+    errors.push('Unable to determine service.')
+    return Promise.reject(errors)
+  }
+
+  if (mergeReqNbr === null || typeof mergeReqNbr === 'undefined') {
+    errors.push('No pull/merge request number found.')
+    return Promise.reject(errors)
+  }
+
+  let review = await gitService.getReview(mergeReqNbr).catch((error) => {
+    const msg = `Failed to retrieve merge request ${mergeReqNbr} - ${error}`
+    console.error(msg)
+    errors.push(msg)
+    return Promise.reject(errors)
+  })
+
+  if (_calcIsMergeRequestStaticmanGenerated(review)) {
+    if (_calcIsMergeRequestAccepted(review)) {
```

Comment:
```suggestion
  if (_calcIsMergeRequestStaticmanGenerated(review) && _calcIsMergeRequestAccepted(review)) {
```
Nesting isn't necessary here